### PR TITLE
feat: add transaction performance metrics

### PR DESCRIPTION
**Pull request by matthewwalsh0**
## **Description**

Add initial custom instrumentation around transactions initiated from a dApp.

Specifically:

- Refactor `trace` utility function to support no callback argument and only start a trace to be manually ended later.
- Add `endTrace` utility function.
- Add `endBackgroundTrace` action to contribute to background metrics from the UI.
- Always trace if in developer build or `METAMASK_DEBUG` is set.
- Pass `trace` callback and `traceContext` to `TransactionController`.
- End `Notification Display` trace (started from controller) as soon as transaction confirmation is rendered with matching ID.

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/26332?quickstart=1)

## **Related issues**

Fixes: [#2850](https://github.com/MetaMask/MetaMask-planning/issues/2850)

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Comment by matthewwalsh0** [comment]
@metamaskbot update-policies

**Review by jpuri — DISMISSED** [review]
Changes look good 👍

**Review by pedronfigueiredo — DISMISSED** [review]

File: app/scripts/lib/createTracingMiddleware.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    const { id, method } = req;
+
+    if (method === MESSAGE_TYPE.ETH_SEND_TRANSACTION) {
+      req.traceContext = await trace({
+        name: 'Transaction',
+        id,
+        tags: { source: 'dapp' },
+      });
+
+      await trace({ name: 'Middleware', id, parentContext: req.traceContext });
```
Comment (davidmurdoch) [inline app/scripts/lib/createTracingMiddleware.ts:22]: should this be in a try/catch?

Comment (matthewwalsh0) [reply app/scripts/lib/createTracingMiddleware.ts:22]: I don't believe there is any known error scenarios here, so it shouldn't introduce any more risk than any other middleware for example.

**Review by HowardBraham — APPROVED** [review]

**Review by pedronfigueiredo — APPROVED** [review]

**Pull request merged**